### PR TITLE
feat(snapshots):  add garbage collectors for pruning and orphaned pending creation resources

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**

### CODE REVIEW

**Comment by Abhinandan-Purkait** [comment]
bors try

**Comment by Abhinandan-Purkait** [comment]
bors try

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs [hunk 1]
```diff
@@ -398,3 +424,54 @@ impl OperationGuardArc<VolumeSnapshot> {
...
+
+async fn check_nodes_availability(
+    replica_snaps: Vec<ReplicaSnapshot>,
+    registry: &Registry,
+) -> bool {
+    let mut nodes: Vec<NodeId> = Vec::new();
+    // Get the list of nodes from the list of replica_snaps.
+    for snap in replica_snaps {
+        if let Ok(pool_spec) = registry.specs().pool(snap.spec().source_id().pool_id()) {
+            nodes.push(pool_spec.node);
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: should this be a hasheset ?

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs [hunk 2]
```diff
@@ -398,3 +424,54 @@ impl OperationGuardArc<VolumeSnapshot> {
         Ok(())
     }
 }
+
+async fn destroy_helper(
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: can we rename this to something else, and comment what it's doing?

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: Ahh right. Missed this. Thanks

File: control-plane/agents/src/bin/core/volume/snapshot_operations.rs (hunk 2 above)
Comment (Abhinandan-Purkait) [reply control-plane/agents/src/bin/core/volume/snapshot_operations.rs]: Done.

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**